### ISSUE TITLE
Memory usage growth in dataset_tools.preprocess

### ISSUE THREAD

**Issue by nsmith-** [scikit-hep/coffea#1007]
**Describe the bug**
Using a memory benchmark on a dataset of about 70 total files (10 files each from 7 samples), I see a stark difference in the memory usage when `save_form` is true:
![image](https://github.com/CoffeaTeam/coffea/assets/6587412/0b68fd1b-d1db-4c0b-957b-b5e32de338c2)

**To Reproduce**
Find some files and make a `samples`, then run on a dask.distributed cluster:
```python
import dask
from coffea import dataset_tools
from lpcjobqueue import LPCCondorCluster
from distributed.diagnostics import MemorySampler

cluster = LPCCondorCluster(ship_env=True)
client = cluster.get_client()

samples = ...

cluster.scale(5)
cluster.wait_for_workers(5)
ms = MemorySampler()
with ms.sample("no form"):
    samples_ready, samples = dataset_tools.preprocess(
        dataset_tools.max_files(samples, 10),
        maybe_step_size=40000,
        skip_bad_files=True,
        recalculate_seen_steps=True,
        save_form=False,
    )
client.restart()
with ms.sample("with form"):
    samples_ready, samples = dataset_tools.preprocess(
        dataset_tools.max_files(samples, 10),
        maybe_step_size=40000,
        skip_bad_files=True,
        recalculate_seen_steps=True,
        save_form=True,
    )
cluster.scale(0)
ms.plot(align=True)
```

**Expected behavior**
Less memory usage

**Comment by lgray** [comment]
That's interesting, it's only like 9kb per file when the form is compressed and encoded...

**Comment by lgray** [comment]
Oh I know what it is - it's holding everything in memory in some wonky serialized format until it combined at the very end. I can turn this into a tree reduction and it should alleviate this.

**Comment by lgray** [comment]
Will have to think about it a bit.

**Comment by lgray** [comment]
@douglasdavis @agoose77 Any idea what might cause memory blowup like this in dask_awkward for byte arrays?

**Comment by douglasdavis** [comment]
Not off the top of my head. What is `save_form` doing here? Any way this can be broken down into an ak+dak only reproducer?

**Comment by lgray** [comment]
Found it. The memory leak appears to be in `uproot.dask`.

@jpivarski Trying to make a reproducer.

**Comment by lgray** [comment]
I've found a workaround in the mean time. Cool.

**Comment by lgray** [comment]
This is an actual memory leak in `uproot.dask` see https://github.com/scikit-hep/uproot5/issues/1093 for a local reproducer.

**Comment by lgray** [comment]
Closed by #1008
